libraries/ms-common/src/stm32f0xx/spi.c

```diff
@@ -0,0 +1,150 @@
+#pragma once
+
+#include "spi.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "gpio.h"
+#include "status.h"
+#include "stm32f0xx.h"
+#include "stm32f0xx_rcc.h"
+#include "stm32f0xx_spi.h"
+
+static SPI_TypeDef *spi_periph_map[] = { SPI1, SPI2 };
+
+static bool prv_is_periph_valid(const uint8_t spi_x) {
+	return spi_x == 0 || spi_x = 1;
+}
+
+static bool prv_are_settings_valid(const SPISettings *settings) {
+	return !(settings->polarity > SPI_CPOLARITY_HIGH ||
+           settings->phase > SPI_CPHASE_2EDGE ||
+           settings->baud_rate > SPI_BAUDRATE_256 ||
+           settings->first_bit > SPI_FIRSTBIT_LSB);
+}
+
+static uint8_t prv_exchange(uint8_t spi_x, uint8_t data) {
+  while(!SPI_I2S_GetFlagStatus(SPI_I2S_FLAG_TXE));
+  SPI_SendData8(spi_periph_map[spi_x], data);
+
+  while(!SPI_I2S_GetFlagStatus(SPI_I2S_FLAG_RXNE));
+  return SPI_ReceiveData8(spi_periph_map[spi_x]);
+}
+
+StatusCode spi_init(uint8_t spi_x, SPISettings *settings) {
+  if (!prv_is_periph_valid(spi_x) || !prv_are_settings_valid(settings)) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  // Enable peripheral clock
+  if (spi_x == 1) {
+    RCC_APB2PeriphClockCmd(RCC_APB2Periph_SPI1, ENABLE);
+  } else {
+    RCC_APB1PeriphClockCmd(RCC_APB1Periph_SPI2, ENABLE);
+  }
+
+  // Note: gpio_init_pin initializes pin clocks as well
+
+  // Initialize MOSI pin
+  GPIOSettings pin_settings;
```

Comment:
I would take advantage of compound structs and/or take advantage of partial initialization. http://stackoverflow.com/questions/11152160/initializing-a-struct-to-0
Otherwise, the variable will actually contain random data.